cpp/references/data_structures/list_digraph.hpp

```diff
@@ -0,0 +1,169 @@
+#ifndef MY_CLASS_HPP
+#define MY_CLASS_HPP
+
+#include <Godot.hpp>
+#include <Reference.hpp>
+
+#include <lemon/list_graph.h>
+#include <lemon/concepts/maps.h>
+
+#include "data_owner.hpp"
+
+namespace godot {
+
+#ifndef ERR_CONTINUE
+#define ERR_CONTINUE(a)   	\
+		if (a) {           	\
+			WARN_PRINT(#a); \
+			continue;      	\
+		}                  	
+#endif
+
+template <class T>
+struct DataWrapper {
+	Variant data = Variant();
+	T* wrapper;
+};
+
+typedef lemon::ListDigraph LemonGraph;
+typedef lemon::ListDigraph::Arc LemonArc;
+typedef lemon::ListDigraph::Node LemonNode;
+typedef lemon::ListDigraph::NodeIt LemonNodeIt;
+
+template<typename T> using LemonArcMap = lemon::ListDigraph::ArcMap<DataWrapper<T>>;
+template<typename T> using LemonNodeMap = lemon::ListDigraph::NodeMap<DataWrapper<T>>;
+
+class ListDigraph : public Reference {
+	GODOT_CLASS(ListDigraph, Reference)
+
+	public:
+		class Vertex : public DataOwner<ListDigraph, Vertex, LemonNode, LemonNodeMap> {
+			GODOT_SUBCLASS(Vertex, DataOwner);
+
+			public:
+				LemonNodeMap<Vertex> *get_map() const override;
+
+				void _init();
+
+				static void _register_methods();
+
+				Vertex();
+				~Vertex();
+
+		};
+
+		class Arc : public DataOwner<ListDigraph, Arc, LemonArc, LemonArcMap> {
+			GODOT_SUBCLASS(Arc, DataOwner);
+
+			public:
+				LemonArcMap<Arc> *get_map() const override;
+
+				void reverse();
+
+				Ref<Vertex> split();
+
+				Ref<Vertex> get_source() const;
+
+				Ref<Vertex> get_target() const;
+
+				void _set_source(Variant p_vertex);
+				void set_source(Ref<Vertex> p_vertex);
+
+				void _set_target(Variant p_vertex);
+				void set_target(Ref<Vertex> p_vertex);
+
+				static void _register_methods();
+
+				void _init();
+
+				Arc();
+				~Arc();
+		};
+
+	private:
+		LemonGraph _graph;
+		LemonNodeIt _vertex_it;
+
+		LemonNodeMap<Vertex> *_vertex_map;
+		LemonArcMap<Arc> *_arc_map;
+
+	public:
+		LemonGraph *get_graph();
+		LemonNodeMap<Vertex> *get_vertex_map();
+		LemonArcMap<Arc> *get_arc_map();
+
+		Array add_vertices(Array p_data, bool p_return);
+
+		Ref<Arc> add_arc(Ref<Vertex> p_source, Ref<Vertex> p_target);
+
+		Array add_arcs(Array p_data, bool p_return);
+
+		void erase_vertex(Ref<Vertex> p_vertex);
+
+		void erase_vertices(Array p_vertices);
+
+		void erase_arc(Ref<Arc> p_arc);
+
+		void erase_arcs(Array p_arcs);
+
+		void contract(Ref<Vertex> p_vertex1, Ref<Vertex> p_vertex2, bool p_remove);
+
+		void reserve_vertex(int p_amount);
+
+		void reserve_arc(int p_amount);
+
+		Ref<Vertex> get_vertex(int p_id);
+
+		Ref<Arc> get_arc(int p_id);
+
+		/* ##### GDSCRIPT API ##### */
+
+		bool _iter_init();
+
+		bool _iter_next();
+
+		Ref<Vertex> _iter_get();
+
+		Ref<Vertex> add_vertex();
+
+		void clear();
+
+		Array get_vertices();
+
+		Array get_arcs();
+
+		Array _add_vertices(Variant p_data, Variant p_return);
```

Comment:
The methods which simply exist to cooperate with scripted bindings should be put under "protected" rather than "public". They should *still* work, but if not, then keep them public out of necessity and mark them separately with comments, as you've done with "GDSCRIPT API" comments.

Reply:
This includes the iteration methods.